grpc/src/main/java/com/linecorp/armeria/common/grpc/GsonGrpcJsonMarshallerBuilder.java

```diff
@@ -32,40 +33,98 @@
 public final class GsonGrpcJsonMarshallerBuilder {
 
     @Nullable
-    private Consumer<JsonFormat.Parser> jsonParserCustomizer;
+    private Function<JsonFormat.Parser, JsonFormat.Parser> jsonParserCustomizer;
 
     @Nullable
-    private Consumer<JsonFormat.Printer> jsonPrinterCustomizer;
+    private Function<JsonFormat.Printer, JsonFormat.Printer> jsonPrinterCustomizer;
 
     GsonGrpcJsonMarshallerBuilder() {}
 
+    /**
+     * Adds a {@link Function} that returns customized the {@link JsonFormat.Parser}
+     * used when deserializing a JSON payload into a {@link Message}.
+     */
+    public GsonGrpcJsonMarshallerBuilder jsonParserCustomizer(
+            Function<? super JsonFormat.Parser, JsonFormat.Parser> jsonParserCustomizer) {
+        requireNonNull(jsonParserCustomizer, "jsonParserCustomizer");
+        if (this.jsonParserCustomizer == null) {
+            @SuppressWarnings("unchecked")
+            final Function<JsonFormat.Parser, JsonFormat.Parser> cast =
+                    (Function<JsonFormat.Parser, JsonFormat.Parser>) jsonParserCustomizer;
+            this.jsonParserCustomizer = cast;
+        } else {
+            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(jsonParserCustomizer);
+        }
+        return this;
+    }
+
     /**
      * Adds a {@link Consumer} that can customize the {@link JsonFormat.Parser}
      * used when deserializing a JSON payload into a {@link Message}.
+     *
+     * @deprecated {@link JsonFormat.Parser} is immutable so all changes applied in the {@link Consumer}
+     *     will be lost. Please use the {@link #jsonParserCustomizer(Function) jsonParserCustomizer}
+     *     which accepts {@link Function} parameter instead.
      */
+    @Deprecated
     public GsonGrpcJsonMarshallerBuilder jsonParserCustomizer(
             Consumer<? super JsonFormat.Parser> jsonParserCustomizer) {
         requireNonNull(jsonParserCustomizer, "jsonParserCustomizer");
         if (this.jsonParserCustomizer == null) {
             @SuppressWarnings("unchecked")
             final Consumer<JsonFormat.Parser> cast = (Consumer<JsonFormat.Parser>) jsonParserCustomizer;
-            this.jsonParserCustomizer = cast;
+            this.jsonParserCustomizer = parser -> {
+                cast.accept(parser);
+                return parser;
+            };
         } else {
-            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(jsonParserCustomizer);
+            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(parser -> {
+                jsonParserCustomizer.accept(parser);
+                return parser;
+            });
         }
         return this;
```

Comment:
How about logging a warning message when this method is called so that people can avoid this pitfall at all? e.g.

```suggestion
        if (!loggedJsonParserCustomizerWarning) {
            logger.warn("{}.jsonParserCustomizer(Consumer) is buggy; use jsonParserCustomizer(Function).",
                        getClass().getSimpleName());
            loggedJsonParserCustomizerWarning = true;
        }
        return this;
```

(`loggedJsonParserCustomizerWarning` could be `private static boolean`.)

Reply:
good idea, added logging https://github.com/line/armeria/pull/6146/commits/0812f5a2c828b8f5878b40d2c5a3b6e93a86ddee